### PR TITLE
Add structure validation for external matrix

### LINKED ISSUE #9: Add structure validation for matrix on gameboard
Currently the _validate_matrix method only checks if the values on the input matrix are valid.
Need to add structure validation to check if the distribution of the values on the input matrix is accordding to the rules of the game. i.e: Check that there are no occupied cells above an empty one.

### PR DESCRIPTION
**Pull request by apojomovsky**
- Fixes #9 
- Also redefines implementation of the __ne__ method as the negation of __eq__
- __eq__ doesn't break if compared objects are not of the same type

### CODE REVIEW

File: game/gameboard.py [hunk 1]
```diff
@@ -149,7 +146,7 @@ def set_board_from_matrix(self, matrix):
             a boolean value
         """
         numpy_matrix = numpy.asarray(matrix)
-        if self._validate_matrix(numpy_matrix):
```
Comment (basicNew) [inline game/gameboard.py:152]: Hmm, why is this method no longer private?

File: game/gameboard.py [hunk 2]
```diff
@@ -249,7 +248,12 @@ def _validate_matrix(self, matrix_to_test):
                     pass
```
Comment (basicNew) [inline game/gameboard.py]: The `if entry in valid_entries` part could be decoupled into a `_is_valid_cell_value(entry)` method. You can also change the code to be:

```
if not self._is_valid_cell_value(entry)
    return False
```

File: game/gameboard.py [hunk 3]
```diff
@@ -249,7 +248,12 @@ def _validate_matrix(self, matrix_to_test):
...
                 else:
                     return False
-        # TODO: Add structure validation (no "floating" entries in matrix)
+        for i in range(7):
```
Comment (basicNew) [inline game/gameboard.py]: Please, use more meaningful names than `i` and `x`

**Review by basicNew — COMMENTED** [review]
Left a couple of comments

File: game/gameboard.py (hunk 1 above)
Comment (apojomovsky) [reply game/gameboard.py:152]: I wanted to write some tests for this method, but figured out that the python linter complained about me trying to call private methods on the test, so I made it public for a while and then I think I forgot to return it to its previous condition. Is that OK if I want to write tests for private methods? (It was really helpful) I've read that it's supposed to be tested indirectly from the non-private methods' tests

**Comment by apojomovsky** [comment]
Thank you for reviewing this, @basicNew. I just addressed all your comments and updated the PR

File: game/gameboard.py [hunk 4]
```diff
@@ -228,30 +225,39 @@ def _check_columns_for_winner(self):
...
+            matrix_to_test: list
         Returns
             a boolean value
         """
-        valid_entries = (None, 'blue', 'red')
+        matrix_to_test = numpy.asarray(matrix_to_test)
         if matrix_to_test.shape != (6,7):
             return False
         for row in matrix_to_test:
             for entry in row:
```
Comment (basicNew) [inline game/gameboard.py]: You can improve this loop with the use of `any()`. See https://docs.python.org/3/library/functions.html#any and http://stackoverflow.com/questions/1342601/pythonic-way-of-checking-if-a-condition-holds-for-any-element-of-a-list

**Review by basicNew — COMMENTED** [review]
Private methods shouldn't be tested directly from a unit test. Please, add this as a topic for our chat on Wed.

File: game/gameboard.py [hunk 5]
```diff
@@ -242,9 +242,8 @@ def _validate_matrix(self, matrix_to_test):
         if matrix_to_test.shape != (6,7):
             return False
         for row in matrix_to_test:
-            for entry in row:
-                if not self._is_valid_cell_value(entry):
-                    return False
+            if not all([self._is_valid_cell_value(entry) for entry in row]):
```
Comment (basicNew) [inline game/gameboard.py]: No need for list comprehension :) http://stackoverflow.com/questions/405516/if-all-in-list-something

Comment (apojomovsky) [reply game/gameboard.py]: great! thanks!

**Comment by apojomovsky** [comment]
Addressed the last comment on the last update, I'm merging this PR now

**Pull request merged**